util.py

```diff
@@ -250,8 +251,22 @@ def fetchFile(source, dest):
                source[:6] == 'https:' or \
                source[:5] == 'file:' or \
                source[:4] == 'ftp:':
-            # This something that can be fetched using urllib2:
-            fd = urllib2.urlopen(source)
+            # This something that can be fetched using urllib2
+            request = source
+
+            if source[:5] == 'http:' and '@' in source: # Auth HTTP
+                pattern = "http://(.+:.+)@(.+)"
```

Comment:
Can we use urlparse here to extract the username and password rather than a regex? (https://docs.python.org/2/library/urlparse.html?highlight=urlparse#urlparse.urlparse)

Reply:
Here is the constructor from python-libs: xcp.accessor.HTTPAccessor() which is tested with Python2 and Python3:
https://github.com/xenserver/python-libs/blob/master/xcp/accessor.py#L366C1-L383C55

This test is unit-tested with Python2 and Python3 and should support also urlencoded Unicode strings for username and password (e.g. "Tanım" which is encoded as "Tan%u0131m") due to the use of urllib.parse.unquote() to get the unquoted string.

The urllib parser works with HTTPS and is unit-tested with HTTP basic auth using a simple ASCII username and password, so that's a good reminder to improve the test to use urlencoded username and password instead.

If the answerfile is red as Python2 string (bytes), and urllib just assumes UTF-8 bytes, not using unquote might work, but it should be better to use urllib.parse.unquote() to support such urlencoded username/password just in case it helps with the whole jungle of different charset encodings.

Reply:
The PR to test HTTP Basic Auth with user and password using URL-encoded Unicode is submitted:
https://github.com/xenserver/python-libs/pull/117

Reply:
I'll reuse the URL class defined in this file, it does URL parsing and extracts the username/password for me :)